### ISSUE TITLE
for CNN in fp16 execution time depends on input scale

### ISSUE THREAD

**Issue by snailrowen1337** [pytorch/pytorch#50354]
## 🐛 Bug

Using a convolutional model in half-precision has execution time dependent upon the input scale.

## To Reproduce

```
import torch
import torch.nn as nn
import torch.nn.functional as F



class network(nn.Module):
    def __init__(self):
        super().__init__()
        self.num_layers = 4
        self.num_filters = 32
        self.convs = nn.ModuleList([
            nn.Conv2d(9, self.num_filters, 3, stride=2),
            nn.Conv2d(self.num_filters, self.num_filters, 3, stride=1),
            nn.Conv2d(self.num_filters, self.num_filters, 3, stride=1),
            nn.Conv2d(self.num_filters, self.num_filters, 3, stride=1)
        ])

    def forward(self, obs):
        conv = torch.relu(self.convs[0](obs))
        for i in range(1, self.num_layers):
            conv = torch.relu(self.convs[i](conv))
        h = conv.view(conv.size(0), -1)
        return h


def time(model, x):
    torch.cuda.synchronize()
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    for _ in range(100):
        y = model(x)
        loss = (y**2).sum()
        loss.backward()
    end.record()
    torch.cuda.synchronize()
    return start.elapsed_time(end)


torch.manual_seed(0)

# fp16
x = torch.randn(512, 9, 84, 84).half().cuda()
model = network().half().cuda()
t1 = time(model, x)
x = x*1e5
t2 = time(model, x)
print('half precision:', 'time on randn', t1, 'time on randn*1e5', t2)


# fp32
x = torch.randn(512, 9, 84, 84).cuda()
model = network().cuda()
t1 = time(model, x)
x = x*1e5
t2 = time(model, x)
print('full precision:', 'time on randn', t1, 'time on randn*1e5', t2)

```

the output is

```
half precision: time on randn 2078.74755859375 time on randn*1e5 1798.8956298828125
full precision: time on randn 3340.018798828125 time on randn*1e5 3340.420166015625
```

## Expected behavior

The time shouldn't be dependent upon the scale of the input.

## Environment

 - PyTorch Version (e.g., 1.0): 1.7.1.
 - OS (e.g., Linux): CentOS
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): n/a
 - Python version: 3.7
 - CUDA/cuDNN version: 10.2/ 7.6.0.5
 - GPU models and configuration: Tesla v100

**Comment by goodclass** [comment]
```
half precision: time on randn 6451.7368164062 time on randn*1e5 5320.66699218
half precision: time on randn 5445.7333984375 time on randn*1e5 5334.55859375
full precision: time on randn 6019.4536132812 time on randn*1e5 5958.53515625
```

- PyTorch Version (e.g., 1.0): 1.7.1
- OS Win10
- GPU models and configuration: RTX 3090
- CUDA/cuDNN version: 11.0/ 8.0.4

**Comment by ngimel** [comment]
cc @ptrblck

**Comment by snailrowen1337** [comment]
This issue persists if I "warm-start" the GPU:

```
def time(model, x):
    for _ in range(100):
        y = model(x)
        loss = (y**2).sum()
        loss.backward()
    torch.cuda.synchronize()
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    for _ in range(100):
        y = model(x)
        loss = (y**2).sum()
        loss.backward()
    end.record()
    torch.cuda.synchronize()
    return start.elapsed_time(end)
```

**Comment by ptrblck** [comment]
I cannot reproduce this issue after adding the warmup iterations via the for loop or by calling `time(model, x)` once before the profiling (I would nevertheless recommend to use a few iterations for warmup and the actual profiling).
On a TitanV I get using the `1.7.1` binaries with CUDA11.0:
Without warmup:
```python
half precision: time on randn 2831.64892578125 time on randn*1e5 2221.00244140625
full precision: time on randn 2623.826171875 time on randn*1e5 2621.57958984375

# swap the calls
full precision: time on randn 3246.992431640625 time on randn*1e5 2624.52490234375
half precision: time on randn 2227.943359375 time on randn*1e5 2228.245849609375
```
With warmup:
```python
half precision: time on randn 2233.7626953125 time on randn*1e5 2229.6318359375
full precision: time on randn 2621.433837890625 time on randn*1e5 2623.6787109375

# swap the calls
full precision: time on randn 2622.098876953125 time on randn*1e5 2622.799560546875
half precision: time on randn 2224.698486328125 time on randn*1e5 2228.815185546875
```

**Comment by ngimel** [comment]
Please use torch Timer utilities with blocked_autorange, they take care of warmup and synchronization https://pytorch.org/docs/master/benchmark_utils.html?highlight=timer#torch.utils.benchmark.Timer

**Comment by snailrowen1337** [comment]
I've tried this and still get inconsistent results based upon the scale:

```
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.utils.benchmark as benchmark

class network(nn.Module):
    def __init__(self):
        super().__init__()
        self.num_layers = 4
        self.num_filters = 32
        self.convs = nn.ModuleList([
            nn.Conv2d(9, self.num_filters, 3, stride=2),
            nn.Conv2d(self.num_filters, self.num_filters, 3, stride=1),
            nn.Conv2d(self.num_filters, self.num_filters, 3, stride=1),
            nn.Conv2d(self.num_filters, self.num_filters, 3, stride=1)
        ])

    def forward(self, obs):
        conv = torch.relu(self.convs[0](obs))
        for i in range(1, self.num_layers):
            conv = torch.relu(self.convs[i](conv))
        h = conv.view(conv.size(0), -1)
        return h

def my_func(model, x):
    y = model(x)
    loss = (y**2).sum()
    loss.backward()


print('using cudnn:', torch.backends.cudnn.version())
torch.manual_seed(0)

# fp16
x = torch.randn(512, 9, 84, 84).half().cuda()
model = network().half().cuda()
t0 = benchmark.Timer(
    stmt='my_func(model, x)',
    setup='from __main__ import my_func',
    globals={'x': x, 'model' : model})
x = x*1e5
t1 = benchmark.Timer(
    stmt='my_func(model, x)',
    setup='from __main__ import my_func',
    globals={'x': x, 'model' : model})

print('half precision:', 'time on randn', t0.blocked_autorange(), 'time on randn*1e5', t1.blocked_autorange())

print()

# fp32
x = torch.randn(512, 9, 84, 84).cuda()
model = network().cuda()
t0 = benchmark.Timer(
    stmt='my_func(model, x)',
    setup='from __main__ import my_func',
    globals={'x': x, 'model' : model})
x = x*1e5
t1 = benchmark.Timer(
    stmt='my_func(model, x)',
    setup='from __main__ import my_func',
    globals={'x': x, 'model' : model})

print('full precision:', 'time on randn', t0.blocked_autorange(), 'time on randn*1e5', t1.blocked_autorange())
```

gives me:

```
using cudnn: 7605
half precision: time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7fc46653e0f0>
my_func(model, x)
  20.63 ms
  1 measurement, 10 runs , 1 thread time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7fc46653e198>
my_func(model, x)
  Median: 17.97 ms
  2 measurements, 10 runs per measurement, 1 thread

full precision: time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7fc46e411d68>
my_func(model, x)
  33.34 ms
  1 measurement, 10 runs , 1 thread time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7fc4cc796c88>
my_func(model, x)
  33.35 ms
  1 measurement, 10 runs , 1 thread
```

any ideas what might cause this? I installed pytorch from scratch with

```
conda install pytorch torchvision torchaudio cudatoolkit=11.0 -c pytorch
```

**Comment by ptrblck** [comment]
@snailrowen1337 could you swap the FP16 and FP32 runs and post the results here to check, if the very first run still shows a slowdown?

**Comment by snailrowen1337** [comment]
I still get a discrepency after reordering:


```

using cudnn: 7605
full precision: time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7ff64bac7588>
my_func(model, x)
  33.82 ms
  1 measurement, 10 runs , 1 thread time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7ff64bac7630>
my_func(model, x)
  33.89 ms
  1 measurement, 10 runs , 1 thread

half precision: time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7ff64bab9208>
my_func(model, x)
  20.83 ms
  1 measurement, 10 runs , 1 thread time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7ff6586315f8>
my_func(model, x)
  Median: 18.19 ms
  2 measurements, 10 runs per measurement, 1 thread

```

**Comment by ptrblck** [comment]
Thanks for the update. Is the "unscaled" (`randn` input) FP16 test consistently returning the larger runtime?
I.e. could you repeat this test a couple of times and swap the order of the raw input vs. scaled input?

**Comment by snailrowen1337** [comment]
yes!

```

for _ in range(3):

    # fp16
    x = torch.randn(512, 9, 84, 84).half().cuda()
    model = network().half().cuda()
    t0 = benchmark.Timer(
        stmt='my_func(model, x)',
        setup='from __main__ import my_func',
        globals={'x': x, 'model' : model})
    xs = x*1e5
    t1 = benchmark.Timer(
        stmt='my_func(model, x)',
        setup='from __main__ import my_func',
        globals={'x': xs, 'model' : model})
    print('half precision:', 'time on randn', t0.blocked_autorange(), 'time on randn*1e5', t1.blocked_autorange())


    # fp16 switch order
    x = torch.randn(512, 9, 84, 84).half().cuda()
    model = network().half().cuda()
    t0 = benchmark.Timer(
        stmt='my_func(model, x)',
        setup='from __main__ import my_func',
        globals={'x': x, 'model' : model})
    xs = x*1e5
    t1 = benchmark.Timer(
        stmt='my_func(model, x)',
        setup='from __main__ import my_func',
        globals={'x': xs, 'model' : model})
    print('half precision:', 'time on randn*1e5', t1.blocked_autorange(), 'time on randn', t0.blocked_autorange())

    print()
```
gives me

```
half precision: time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcaf3240>
my_func(model, x)
  20.78 ms
  1 measurement, 10 runs , 1 thread time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0e9669630>
my_func(model, x)
  Median: 18.13 ms
  2 measurements, 10 runs per measurement, 1 thread
half precision: time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcb00208>
my_func(model, x)
  Median: 18.13 ms
  2 measurements, 10 runs per measurement, 1 thread time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcb00470>
my_func(model, x)
  20.72 ms
  1 measurement, 10 runs , 1 thread

half precision: time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcaf3240>
my_func(model, x)
  20.66 ms
  1 measurement, 10 runs , 1 thread time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7fc13b1961d0>
my_func(model, x)
  Median: 18.13 ms
  2 measurements, 10 runs per measurement, 1 thread
half precision: time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcb00588>
my_func(model, x)
  Median: 18.12 ms
  2 measurements, 10 runs per measurement, 1 thread time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcb002b0>
my_func(model, x)
  20.82 ms
  1 measurement, 10 runs , 1 thread

half precision: time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcaf3240>
my_func(model, x)
  20.88 ms
  1 measurement, 10 runs , 1 thread time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0e9669630>
my_func(model, x)
  Median: 18.10 ms
  2 measurements, 10 runs per measurement, 1 thread
half precision: time on randn*1e5 <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcb00908>
my_func(model, x)
  Median: 18.14 ms
  2 measurements, 10 runs per measurement, 1 thread time on randn <torch.utils.benchmark.utils.common.Measurement object at 0x7fc0dcb00518>
my_func(model, x)
  20.75 ms
  1 measurement, 10 runs , 1 thread


```

**Comment by snailrowen1337** [comment]
the discrepancy almost disappears if I set

```
torch.set_deterministic(True)
torch.backends.cudnn.deterministic = True
```

Most likely this is a matter of the algorithm/kernel selection of the cudnn/cuda runtime which probably differs between GPU models. Feel free to close this issue.

**Comment by ptrblck** [comment]
@snailrowen1337 thanks for the follow-up. I can reproduce it using cudnn7.5.6 on a V100 and think the slowdown might trigger slower code paths to handle denormal values. I'll check with our cudnn team, if this could be the issue.
As said before, cudnn8 does not show this behavior, so you might want to install the binaries with the CUDA11.0 runtime.

Note that you are also creating overflows by scaling the `FP16` inputs with such a high value (I assume this scaling factor is not really used in your code, just for this benchmark).

**Comment by snailrowen1337** [comment]
I get this result even if I do

```
x = torch.randn(BATCH_S, 3, SHAPE, SHAPE, dtype=torch.float16).cuda()
```

I assume that this should not create any denormal values?
